Adding ANSI_WARNINGS explainer, re: COUNT is int NULL (#8059)
MicrosoftDocs/sql-docs: docs/t-sql/functions/count-transact-sql.md

The *partition_by_clause* divides the result set produced by the `FROM` clause into partitions to which the `COUNT` function is applied. If not specified, the function treats all rows of the query result set as a single group. The *order_by_clause* determines the logical order of the operation. See [OVER Clause &#40;Transact-SQL&#41;](../../t-sql/queries/select-over-clause-transact-sql.md) for more information. 

## Return types
 

>  * **int NOT NULL** when `ANSI_WARNINGS` is `ON`, however SQL Server will always treat `COUNT` expressions as `int NULL` in metadata, unless wrapped in `ISNULL`.

Reviewer: ```suggestion
- **int NOT NULL** when `ANSI_WARNINGS` is `ON`, however SQL Server will always treat `COUNT` expressions as `int NULL` in metadata, unless wrapped in `ISNULL`.
```

>  * **int NULL** when `ANSI_WARNINGS` is `OFF`.

Reviewer: ```suggestion

- **int NULL** when `ANSI_WARNINGS` is `OFF`.
```

  
## Remarks  

> * `COUNT(\*)` without `GROUP BY` returns the cardinality (number of rows) in the resultset. This includes rows comprised of all-`NULL` values and duplicates.

Reviewer: ```suggestion
* `COUNT(*)` without `GROUP BY` returns the cardinality (number of rows) in the resultset. This includes rows comprised of all-`NULL` values and duplicates.

```

> * `COUNT(\*)` with `GROUP BY` returns the number of rows in each group. This includes `NULL` values and duplicates.

Reviewer: ```suggestion
* `COUNT(*)` with `GROUP BY` returns the number of rows in each group. This includes `NULL` values and duplicates.
```

> * `COUNT(ALL *expression*)` evaluates *expression* for each row in a group, and returns the number of nonnull values.

Reviewer: ```suggestion
* `COUNT(ALL <expression>)` evaluates *expression* for each row in a group, and returns the number of nonnull values.
```

* `COUNT(DISTINCT *expression*)` evaluates *expression* for each row in a group, and returns the number of unique, nonnull values.
  
`COUNT` is a deterministic function when used ***without*** the OVER and ORDER BY clauses. It is nondeterministic when used ***with*** the OVER and ORDER BY clauses. See [Deterministic and Nondeterministic Functions](../../relational-databases/user-defined-functions/deterministic-and-nondeterministic-functions.md) for more information.

### `ARITHABORT` and `ANSI_WARNINGS`

> When `COUNT` will have a return value exceeding the maximum value of `int` (i.e. 2<sup>31</sup>-1 or 2,147,483,647) then the `COUNT` function will fail due to integer overflow. When `COUNT` overflows and _both_ the `ARITHABORT` and `ANSI_WARNINGS` options are `OFF` then `COUNT` will return `NULL`; otherwise (when _either_ of `ARITHABORT` or `ANSI_WARNINGS` are `ON`) then the query will be aborted and the arithmetic overflow error "Msg 8115, Level 16, State 2; Arithmetic overflow error converting expression to data type int." will be raised. To correctly handle these large results instead use `COUNT_BIG` which returns `bigint`.

Reviewer: ```suggestion
When `COUNT` has a return value exceeding the maximum value of **int** (i.e. 2<sup>31</sup>-1 or 2,147,483,647), the `COUNT` function will fail due to an integer overflow. When `COUNT` overflows and *both* the `ARITHABORT` and `ANSI_WARNINGS` options are `OFF`, `COUNT` will return `NULL`. Otherwise, when *either* of `ARITHABORT` or `ANSI_WARNINGS` are `ON`, the query will abort and the arithmetic overflow error `Msg 8115, Level 16, State 2; Arithmetic overflow error converting expression to data type int.` will be raised. To correctly handle these large results, use `COUNT_BIG` instead, which returns **bigint**.
```

Reviewer 2: You're okay with my wording here? I'm reading it back to myself and it feels like a word-salad.

Reviewer: I was editing for formatting on a weekend and you already called me out on that :). If I were to write it during the week, it would look like this:

```suggestion
If the value of `COUNT` exceeds the maximum value of **int** (2<sup>31</sup>-1 or 2,147,483,647), the function will fail due to an integer overflow. If both `ARITHABORT` and `ANSI_WARNINGS` options are `OFF`, `COUNT` will return `NULL`. If either `ARITHABORT` or `ANSI_WARNINGS` are `ON`, the query will abort and you will receive an arithmetic overflow error. In this scenario, use `COUNT_BIG` instead, which returns **bigint**.
```

> When both ``ARITHABORT`` and `ANSI_WARNINGS` are `ON` you can safely wrap `COUNT` call-sites in `ISNULL( <count-expr>, 0 )` to coerce the expression's type to `int NOT NULL` instead of `int NULL`; otherwise then wrapping `COUNT` is `ISNULL` means any overflow error will be silently suppressed which should be carefully considered for correctness.

Reviewer: ```suggestion
When both ``ARITHABORT`` and `ANSI_WARNINGS` are `ON`, you can safely wrap `COUNT` call-sites in `ISNULL( <count-expr>, 0 )` to coerce the expression's type to `int NOT NULL` instead of `int NULL`. Wrapping `COUNT` in `ISNULL` means any overflow error will be silently suppressed, which should be considered for correctness.
```
